### PR TITLE
Add utility to merge layout segments

### PR DESCRIPTION
**Pull request by lixinchengdu**
## Overview

Description: Add a utility in LayoutBuilder to merge layout segments

Why should this be merged: 

Related issue(s) (if applicable): #1672 

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

**Comment by Maithem** [comment]
Hi @lixinchengdu , why did you close this?

File: runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java [hunk 1]
```diff
@@ -305,6 +315,69 @@ public LayoutBuilder mergePreviousSegment(int segmentIndex) {
...
+     * If an exception happens, leave layout unmodified.
+     * No addition or removal of stripes are allowed.
+     *
+     * @param segmentIndex Segment to merge.
+     * @param allowNonidenticalStripes Allow segments with nonidentical stripes to merge.
+     * @param allowedUpdates Number of allowed server addition/removal between segments, if allow nonidentical stripes.
+     *
+     * @return this builder
+     */
+    public LayoutBuilder attemptMergePreviousSegmentInternal(int segmentIndex, boolean allowNonidenticalStripes,
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java]: More intuitive method name.

File: runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java [hunk 2]
```diff
@@ -272,23 +274,31 @@ public LayoutBuilder mergePreviousSegment(int segmentIndex) {
...
         List<LayoutStripe> oldSegmentStripeList = layoutSegmentList.get(segmentIndex - 1).getStripes();
         List<LayoutStripe> newSegmentStripeList = layoutSegmentList.get(segmentIndex).getStripes();
 
-        int allowedUpdates = 1;
+        int updates = 0;
         if (oldSegmentStripeList.size() != newSegmentStripeList.size()) {
             throw new LayoutModificationException("Stripe addition/deletion not allowed.");
         }
         for (int i = 0; i < oldSegmentStripeList.size(); i++) {
+            if (!allowNonidenticalStripes && !oldSegmentStripeList.get(i).equals(newSegmentStripeList.get(i))) {
```
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java]: This name may be misleading. Two identical stripes means that LogUnitServerList are identical, including the order. 

Existing mergePreviousSegment method belongs to the scenario where two adjacent segments even have different LogUnit Server set.

File: runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java [hunk 3]
```diff
@@ -254,13 +254,15 @@ public LayoutBuilder addLogunitServerToSegment(@NonNull String endpoint,
...
-     * Only 1 log unit server addition/removal allowed between segments.
      *
      * @param segmentIndex Segment to merge.
+     * @param allowNonidenticalStripes Allow segments with nonidentical stripes to merge.
+     * @param allowedUpdates Number of allowed server addition/removal between segments, if allow nonidentical stripes.
      *
      * @return this builder
      */
-    public LayoutBuilder mergePreviousSegment(int segmentIndex) {
+    public LayoutBuilder mergePreviousSegmentInternal(int segmentIndex, boolean allowNonidenticalStripes,
```
[the remark this answers is not in the document]
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java]: Good catch! The helper method should be marked as private.

File: runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java (hunk 1 above)
Comment (lixinchengdu) [reply runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java]: I will discard using this method.

**Comment by Maithem** [comment]
@lixinchengdu can you please rebase. And address comments.

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java  5
         

Complexity decreasing per file
==============================
+ infrastructure/src/main/java/org/corfudb/infrastructure/BatchWriter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/CorfuDB/CorfuDB/pullRequest?prid=4653456&bid=15584347)

**Comment by annym** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 7
           

Complexity increasing per file
==============================
- runtime/src/main/java/org/corfudb/runtime/view/LayoutBuilder.java  5
         

Complexity decreasing per file
==============================
+ infrastructure/src/main/java/org/corfudb/infrastructure/BatchWriter.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651484&bid=15577797)

**Comment by xnull** [comment]
Obsolete

**Pull request closed without merging**